File: src-kotlin/task-manager-infrastructure/src/main/kotlin/com/example/taskmanager/infrastructure/LegacyTaskProcessor.kt
package com.example.taskmanager.infrastructure

/**
 * Legacy workshop code for Lab 3.
 *
 * The code works, but it treats Kotlin like nullable Java and repeats
 * manual null checks instead of using the language's null-safety features.
 */
class LegacyTaskProcessor {
    fun processUpdate(request: LegacyTaskUpdateRequest?): String {
        var summary = ""

        if (request != null) {
            if (request.taskId != null) {
                summary = "Task " + request.taskId
            } else {
                summary = "Task unknown"
            }

            if (request.title != null) {
                if (request.title.trim().isNotEmpty()) {
                    summary = summary + " | title=" + request.title.trim()
                } else {
                    summary = summary + " | title=(blank)"
                }
            } else {
                summary = summary + " | title=(missing)"
            }

            if (request.description != null) {
                if (request.description.trim().isNotEmpty()) {
                    summary = summary + " | description=" + request.description.trim()
                } else {
                    summary = summary + " | description=(blank)"
                }
            } else {
                summary = summary + " | description=(missing)"
            }

            if (request.estimateMinutes != null) {
                if (request.estimateMinutes > 0) {
                    summary = summary + " | estimate=" + request.estimateMinutes + "m"
                } else {
                    summary = summary + " | estimate=unplanned"
                }
            } else {
                summary = summary + " | estimate=unknown"
            }

            if (request.assignee != null) {
                if (request.assignee.trim().isNotEmpty()) {
                    summary = summary + " | assignee=" + request.assignee.trim()
                } else {
                    summary = summary + " | assignee=unassigned"
                }
            } else {
                summary = summary + " | assignee=unassigned"
            }

            if (request.status != null) {
                if (request.status == "DONE") {
                    summary = summary + " | state=complete"
                } else {
                    if (request.status == "IN_PROGRESS") {
                        summary = summary + " | state=active"
                    } else {
                        summary = summary + " | state=open"
                    }
                }
            } else {
                summary = summary + " | state=open"
            }

            if (request.notifyChannel != null) {
                if (request.notifyRecipient != null) {
                    if (request.notifyRecipient.trim().isNotEmpty()) {
                        summary =
                            summary +
                                " | notify=" +
                                request.notifyChannel +
                                ":" +
                                request.notifyRecipient.trim()
                    } else {
                        summary = summary + " | notify=skipped"
                    }
                } else {
                    summary = summary + " | notify=skipped"
                }
            } else {
                summary = summary + " | notify=skipped"
            }
        }

        return summary
    }
}

data class LegacyTaskUpdateRequest(
    val taskId: String?,
    val title: String?,
    val description: String?,
    val estimateMinutes: Int?,
    val assignee: String?,
    val status: String?,
    val notifyChannel: String?,
    val notifyRecipient: String?,
)

File: src-swift/Sources/TaskManagerInfrastructure/LegacyTaskProcessor.swift
import Foundation

/// Legacy workshop code for Lab 3.
///
/// The code works, but it force-unwraps optionals and builds ad-hoc output
/// strings instead of using safer Swift patterns.
public final class LegacyTaskProcessor {
    public init() {}

    public func processUpdate(_ request: LegacyTaskUpdateRequest?) -> String {
        var summary = ""

        if request != nil {
            if request!.taskID != nil {
                summary = "Task \(request!.taskID!)"
            } else {
                summary = "Task unknown"
            }

            if request!.title != nil {
                if request!.title!.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    summary += " | title=(blank)"
                } else {
                    summary += " | title=\(request!.title!.trimmingCharacters(in: .whitespacesAndNewlines))"
                }
            } else {
                summary += " | title=(missing)"
            }

            if request!.description != nil {
                if request!.description!.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    summary += " | description=(blank)"
                } else {
                    summary += " | description=\(request!.description!.trimmingCharacters(in: .whitespacesAndNewlines))"
                }
            } else {
                summary += " | description=(missing)"
            }

            if request!.estimateMinutes != nil {
                if request!.estimateMinutes! > 0 {
                    summary += " | estimate=\(request!.estimateMinutes!)m"
                } else {
                    summary += " | estimate=unplanned"
                }
            } else {
                summary += " | estimate=unknown"
            }

            if request!.assignee != nil {
                if request!.assignee!.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    summary += " | assignee=unassigned"
                } else {
                    summary += " | assignee=\(request!.assignee!.trimmingCharacters(in: .whitespacesAndNewlines))"
                }
            } else {
                summary += " | assignee=unassigned"
            }

            if request!.status != nil {
                if request!.status! == "DONE" {
                    summary += " | state=complete"
                } else if request!.status! == "IN_PROGRESS" {
                    summary += " | state=active"
                } else {
                    summary += " | state=open"
                }
            } else {
                summary += " | state=open"
            }

            if request!.notifyChannel != nil {
                if request!.notifyRecipient != nil {
                    if request!.notifyRecipient!.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                        summary += " | notify=skipped"
                    } else {
                        summary += " | notify=\(request!.notifyChannel!):\(request!.notifyRecipient!.trimmingCharacters(in: .whitespacesAndNewlines))"
                    }
                } else {
                    summary += " | notify=skipped"
                }
            } else {
                summary += " | notify=skipped"
            }
        }

        return summary
    }
}

public struct LegacyTaskUpdateRequest {
    public let taskID: String?
    public let title: String?
    public let description: String?
    public let estimateMinutes: Int?
    public let assignee: String?
    public let status: String?
    public let notifyChannel: String?
    public let notifyRecipient: String?

    public init(
        taskID: String?,
        title: String?,
        description: String?,
        estimateMinutes: Int?,
        assignee: String?,
        status: String?,
        notifyChannel: String?,
        notifyRecipient: String?
    ) {
        self.taskID = taskID
        self.title = title
        self.description = description
        self.estimateMinutes = estimateMinutes
        self.assignee = assignee
        self.status = status
        self.notifyChannel = notifyChannel
        self.notifyRecipient = notifyRecipient
    }
}
